Mathlib/NumberTheory/ModularForms/EisensteinSeries/UniformConvergence.lean

```diff
@@ -0,0 +1,238 @@
+
+lemma r_pos (z : ℍ) : 0 < r z := by
+  simp only [r, lt_min_iff, im_pos, Real.sqrt_pos, r1_pos, and_self]
+
+lemma r1_aux_bound (z : ℍ) (δ : ℝ) {ε : ℝ} (hε : 1 ≤ ε ^ 2) :
+    r1 z ≤ (δ * z.re + ε) ^ 2 + (δ * z.im) ^ 2 := by
+  have H1 : (δ * z.re + ε) ^ 2 + (δ * z.im) ^ 2 =
+    δ ^ 2 * (z.re ^ 2 + z.im ^ 2) + ε * 2 * δ * z.re + ε ^ 2 := by ring
+  have H2 : (δ ^ 2 * (z.re ^ 2 + z.im ^ 2) + ε * 2 * δ * z.re + ε ^ 2) * (z.re ^ 2 + z.im ^ 2)
+    - z.im ^ 2 = (δ * (z.re ^ 2 + z.im ^ 2) + ε * z.re) ^ 2 + (ε ^ 2 - 1) * z.im ^ 2 := by ring
+  rw [r1, H1, div_le_iff (by positivity), ← sub_nonneg, H2]
+  exact add_nonneg (sq_nonneg _) (mul_nonneg (sub_nonneg.mpr hε) (sq_nonneg _))
+
+lemma auxbound1 (z : ℍ) {δ : ℝ} (ε : ℝ) (hδ : 1 ≤ δ ^ 2) : r z ≤ Complex.abs (δ * z + ε) := by
+  rcases z with ⟨z, hz⟩
+  have H1 : z.im ≤ √((δ * z.re + ε) ^ 2 + (δ * z).im ^ 2) := by
+    rw [Real.le_sqrt' hz, im_ofReal_mul, mul_pow]
+    exact (le_mul_of_one_le_left (sq_nonneg _) hδ).trans <| le_add_of_nonneg_left (sq_nonneg _)
+  simpa only [r, abs_apply, normSq_apply, add_re, re_ofReal_mul, coe_re, ← pow_two, add_im, mul_im,
+    coe_im, ofReal_im, zero_mul, add_zero, min_le_iff] using Or.inl H1
+
+lemma auxbound2 (z : ℍ) (δ : ℝ) {ε : ℝ} (hε : 1 ≤ ε ^ 2) : r z ≤ Complex.abs (δ * z + ε) := by
+  have H1 : √(r1 z) ≤ √((δ * z.re + ε) ^ 2 + (δ * z.im) ^ 2) :=
+    (Real.sqrt_le_sqrt_iff (by positivity)).mpr (r1_aux_bound _ _ hε)
+  simpa only [r, abs_apply, normSq_apply, add_re, re_ofReal_mul, coe_re, ofReal_re, ← pow_two,
+    add_im, im_ofReal_mul, coe_im, ofReal_im, add_zero, min_le_iff] using Or.inr H1
+
+lemma left_ne_zero_of_max_eq {x : Fin 2 → ℤ} (hx : x ≠ 0)
+    (h : max (x 0).natAbs (x 1).natAbs = (x 0).natAbs) : x 0 ≠ 0 := by
+  contrapose! hx
+  rw [hx, Int.natAbs_zero, max_eq_left_iff, Nat.le_zero, Int.natAbs_eq_zero] at h
+  ext1 j
+  fin_cases j <;> assumption
+
+lemma right_ne_zero_of_max_eq {x : Fin 2 → ℤ} (hx : x ≠ 0)
+    (h : (max (x 0).natAbs (x 1).natAbs) = (x 1).natAbs) : x 1 ≠ 0 := by
+  contrapose! hx
+  rw [hx, Int.natAbs_zero, max_eq_right_iff, Nat.le_zero, Int.natAbs_eq_zero] at h
+  ext1 j
+  fin_cases j <;> assumption
+
+lemma div_max_sq_ge_one (x : Fin 2 → ℤ) (hx : x ≠ 0) :
+    (1 : ℝ) ≤ (x 0 / (max (x 0).natAbs (x 1).natAbs)) ^ 2 ∨
+      (1 : ℝ) ≤ (x 1 / (max (x 0).natAbs (x 1).natAbs)) ^ 2 := by
+  refine (max_choice (x 0).natAbs (x 1).natAbs).imp (fun H1 => ?_) (fun H2 => ?_)
+  · simp only [Fin.isValue, H1, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
+    OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero,
+    left_ne_zero_of_max_eq hx H1, div_self, le_refl]
+  · simp only [Fin.isValue, H2, Int.cast_natAbs, Int.cast_abs, div_pow, _root_.sq_abs, ne_eq,
+    OfNat.ofNat_ne_zero, not_false_eq_true, pow_eq_zero_iff, Int.cast_eq_zero,
+    right_ne_zero_of_max_eq hx H2, div_self, le_refl]
+
+lemma rpow_bound {k : ℝ} (hk : 0 ≤ k) (z : ℍ) (x : Fin 2 → ℤ) (hx : x ≠ 0) :
```

Comment:
Here you prove an inequality with `k`-th powers; but since it is actually use the `-k`-th powers that you use later, you have to invoke `Real.rpow_neg` three times over and prove a positivity hypothesis each time. Why not just forget `k` in this lemma and prove `r z * max (x 0).natAbs (x 1).natAbs ≤ Complex.abs (x 0 * z + x 1)` (it will need a new name of course)?

Reply:
see comment below.